modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2051,13 +2051,48 @@ def sort_function(df):
         # If this df is empty, we don't want to try and shuffle or sort.
         if len(self.axes[0]) == 0 or len(self.axes[1]) == 0:
             return self.copy()
-        # If this df only has one row partition, we don't want to do a shuffle and sort - we can
-        # just do a full-axis sort.
-        if len(self._partitions) == 1:
-            return self.apply_full_axis(
-                1,
-                sort_function,
-            )
+
+        ideal_num_new_partitions = len(self._partitions)
+        m = len(self.index) / ideal_num_new_partitions
+        sampling_probability = (1 / m) * np.log(
+            ideal_num_new_partitions * len(self.index)
+        )
+        # If this df is overpartitioned, we try to sample each partition with probability
+        # greater than 1, which leads to an error. In this case, we can do one of the following
+        # two things. If there is only enough rows for one partition, and we have only 1 column
+        # partition, we can just combine the overpartitioned df into one partition, and sort that
+        # partition. If there is enough data for more than one partition, we can tell the sorting
+        # algorithm how many partitions we want to end up with, so it samples and finds pivots
+        # according to that.
+        if sampling_probability >= 1:
+            from modin.config import MinPartitionSize
+
+            ideal_num_new_partitions = round(len(self.index) / MinPartitionSize.get())
+            if len(self.index) < MinPartitionSize.get() or ideal_num_new_partitions < 2:
+                modin_frame = self
+                if self._partitions.shape[1] != 1:
+                    # In this case, we have more than one column partition, so we first need
+                    # to create row-wise partitions with all of the columns, so we don't have
+                    # a KeyError when sorting. This method can be slow if we have very very
+                    # many columns.
+                    new_partitions = self._partition_mgr_cls.row_partitions(
+                        self._partitions
+                    )
+                    new_partitions = np.array(
+                        [[partition] for partition in new_partitions]
+                    )
+                    modin_frame = self.__constructor__(
+                        new_partitions,
+                        *self.axes,
+                        self._row_lengths_cache,
+                        [len(self.columns)],
+                        self.dtypes,
+                    )
+                return modin_frame.apply_full_axis(
+                    1,
+                    sort_function,
+                )
```

Comment:
if I understand correctly, we can still end up here when having multiple row partitions, how is then data being sorted globally? Shouldn't we end up with the result when all row parts are sorted independently and not the whole frame.

So, summing up, will this scenario work?
```python
df:
   col_to_sort
0           20
1           10     <-- row_part1
------------------
2            2     <-- row_part2
3            1

df.apply_full_axis(sort_func, axis=1)
As far as I understand the result would be:
   col_to_sort
0           10
1           20     <-- row_part1
------------------
2            1     <-- row_part2
3            2
```

Reply:
This scenario works! In fact, we even added a test for this very scenario as part of this PR - test_join_sort.py::test_sort_overpartitioned_df tests this very scenario, along with 2 others (with comments to explain what is being tested!)

Reply:
I observe a really strange behavior for this test though. It works perfectly fine when running with pytest, however, the last scenario fails if I try to run it as a dedicated script.

To be precise, this exact python script is failing for me:

<details><summary>python script</summary>

```python
from modin.pandas.test.utils import eval_general

import pandas
import numpy as np
import modin.pandas as pd

data = [[1, 2, 3], [4, 5, 6]]
modin_df = pd.concat([pd.DataFrame(row).T for row in data]).reset_index(drop=True)
pandas_df = pandas.DataFrame(data)

eval_general(modin_df, pandas_df, lambda df: df.sort_values(by=0))

# Next we test when the final df will only have 1 row, but starts with multiple column
# partitions.
data = [list(range(100)), list(range(100, 200))]
modin_df = pd.concat([pd.DataFrame(row).T for row in data]).reset_index(drop=True)
pandas_df = pandas.DataFrame(data)

eval_general(modin_df, pandas_df, lambda df: df.sort_values(by=0))

# Next we test when the final df will have multiple row partitions.
data = np.random.choice(650, 650, replace=False).reshape((65, 10))
modin_df = pd.concat([pd.DataFrame(row).T for row in data]).reset_index(drop=True)
pandas_df = pandas.DataFrame(data)

eval_general(modin_df, pandas_df, lambda df: df.sort_values(by=0)) # fails
```

</details>

With the following output:
<details><summary>output</summary>

```
Traceback (most recent call last):
  File "t3.py", line 26, in <module>
    eval_general(modin_df, pandas_df, lambda df: df.sort_values(by=0))
  File "repos/modin/modin/pandas/test/utils.py", line 826, in eval_general
    values = execute_callable(
  File "repos/modin/modin/pandas/test/utils.py", line 805, in execute_callable
    md_result = fn(modin_df, **md_kwargs)
  File "t3.py", line 26, in <lambda>
    eval_general(modin_df, pandas_df, lambda df: df.sort_values(by=0))
  File "repos/modin/modin/logging/logger_decorator.py", line 128, in run_and_log
    return obj(*args, **kwargs)
  File "repos/modin/modin/pandas/base.py", line 2920, in sort_values
    result = self._query_compiler.sort_rows_by_column_values(
  File "repos/modin/modin/logging/logger_decorator.py", line 128, in run_and_log
    return obj(*args, **kwargs)
  File "repos/modin/modin/core/storage_formats/pandas/query_compiler.py", line 3400, in sort_rows_by_column_values
    new_modin_frame = self._modin_frame.sort_by(
  File "repos/modin/modin/logging/logger_decorator.py", line 128, in run_and_log
    return obj(*args, **kwargs)
  File "repos/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 126, in run_f_on_minimally_updated_metadata
    result = f(self, *args, **kwargs)
  File "repos/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 2119, in sort_by
    new_partitions = self._partition_mgr_cls.shuffle_partitions(
  File "repos/modin/modin/logging/logger_decorator.py", line 128, in run_and_log
    return obj(*args, **kwargs)
  File "repos/modin/modin/core/dataframe/pandas/partitioning/partition_manager.py", line 1583, in shuffle_partitions
    pivots = shuffle_functions.pivot_function(samples)
  File "repos/modin/modin/core/dataframe/pandas/dataframe/utils.py", line 70, in pivot_fn
    return pick_pivots_from_samples_for_sort(
  File "repos/modin/modin/core/dataframe/pandas/dataframe/utils.py", line 205, in pick_pivots_from_samples_for_sort
    return _find_quantiles(all_pivots, quantiles, method)
  File "repos/modin/modin/core/dataframe/pandas/dataframe/utils.py", line 114, in _find_quantiles
    return np.quantile(df, quantiles)
  File "<__array_function__ internals>", line 200, in quantile
  File "miniconda3/envs/modinc/lib/python3.8/site-packages/numpy/lib/function_base.py", line 4461, in quantile
    return _quantile_unchecked(
  File "miniconda3/envs/modinc/lib/python3.8/site-packages/numpy/lib/function_base.py", line 4473, in _quantile_unchecked
    return _ureduce(a,
  File "miniconda3/envs/modinc/lib/python3.8/site-packages/numpy/lib/function_base.py", line 3752, in _ureduce
    r = func(a, **kwargs)
  File "miniconda3/envs/modinc/lib/python3.8/site-packages/numpy/lib/function_base.py", line 4639, in _quantile_ureduce_func
    result = _quantile(arr,
  File "miniconda3/envs/modinc/lib/python3.8/site-packages/numpy/lib/function_base.py", line 4750, in _quantile
    previous = np.take(arr, previous_indexes, axis=DATA_AXIS)
  File "<__array_function__ internals>", line 200, in take
  File "miniconda3/envs/modinc/lib/python3.8/site-packages/numpy/core/fromnumeric.py", line 190, in take
    return _wrapfunc(a, 'take', indices, axis=axis, out=out, mode=mode)
  File "miniconda3/envs/modinc/lib/python3.8/site-packages/numpy/core/fromnumeric.py", line 57, in _wrapfunc
    return bound(*args, **kwds)
IndexError: cannot do a non-empty take from an empty axes.
```

</details>

Could you please verify whether it works for you?

Reply:
BTW, the sampling probability appears to be less than 1 for the multi row-partitions case from the test you're referring:
```python
> repos/modin/modin/core/dataframe/pandas/dataframe/dataframe.py(2068)sort_by()
-> if sampling_probability >= 1:
(Pdb) sampling_probability
0.34219579267787864
```
meaning that we do not visit this contradicting if-branch.

If we rewrite the case so it would hit this branch the behavior is incorrect because of the reason I described above:
```python
import modin.config as cfg
cfg.NPartitions.put(4)

import modin.pandas as pd

part1 = pd.DataFrame({"a": [20, 10]})
part2 = pd.DataFrame({"a": [2, 1]})

df = pd.concat([part1, part2])

assert df._query_compiler._modin_frame._partitions.shape == (2, 1)

res = df.sort_values(by="a")
print(res)
# output:
#     a
# 1  10
# 0  20
# 1   1
# 0   2
```

Reply:
Resolved - it's because I was doing `modin_df.apply_full_axis(axis=1)` (so over the columns) instead of `modin_df.apply_full_axis(axis=0)` (so over the rows). I've gone ahead and fixed this, but thank you so much for catching this!

Reply:
For the second script, we now work:
```python
In [1]: import modin.config as cfg
   ...: cfg.NPartitions.put(4)
   ...:
   ...: import modin.pandas as pd
   ...:
   ...: part1 = pd.DataFrame({"a": [20, 10]})
   ...: part2 = pd.DataFrame({"a": [2, 1]})
   ...:
   ...: df = pd.concat([part1, part2])
   ...:
   ...: assert df._query_compiler._modin_frame._partitions.shape == (2, 1)
   ...:
   ...: res = df.sort_values(by="a")
   ...: print(res)
UserWarning: Ray execution environment not yet initialized. Initializing...
To remove this warning, run the following python code before doing dataframe operations:

    import ray
    ray.init(runtime_env={'env_vars': {'__MODIN_AUTOIMPORT_PANDAS__': '1'}})

UserWarning: Distributing <class 'dict'> object. This may take some time.
UserWarning: Distributing <class 'dict'> object. This may take some time.
    a
1   1
0   2
1  10
0  20
```
For the second script, I had to make it so we concatenate partitions beforehand, otherwise we were sampling partitions with only 1 row, which lead to us having empty samples, and erroring out.
```python
In [1]: from modin.pandas.test.utils import eval_general
   ...: from modin.config import NPartitions; NPartitions.put(24)
   ...: import pandas
   ...: import numpy as np
   ...: import modin.pandas as pd
   ...:
   ...:
   ...: # Next we test when the final df will have multiple row partitions.
   ...: data = np.random.choice(650, 650, replace=False).reshape((65, 10))
   ...: modin_df = pd.concat([pd.DataFrame(row).T for row in data]).reset_index(drop=True)
   ...: pandas_df = pandas.DataFrame(data)
   ...:
   ...: eval_general(modin_df, pandas_df, lambda df: df.sort_values(by=0)) # fails
UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
UserWarning: Ray execution environment not yet initialized. Initializing...
To remove this warning, run the following python code before doing dataframe operations:

    import ray
    ray.init(runtime_env={'env_vars': {'__MODIN_AUTOIMPORT_PANDAS__': '1'}})

UserWarning: Distributing <class 'numpy.ndarray'> object. This may take some time.

In [2]: modin_df.sort_values(by=0)
Out[2]:
      0    1    2    3    4    5    6    7    8    9
32    4   34  371  326  610   75  182   28  260  596
19   24   42  174  208  627  593  387  266  615  169
3    35  115  614  452   55  648  499  163  215  240
48   50  585  257  104  358  126  278  180  538  450
53   68  116  490  306   38  478  463  547  410  565
..  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...
38  602   67  154   16  315  528  225   80  284  368
17  621  274  578  289  441  502  600  445   73  494
54  644  481  377  128  348  227   15  271  641  524
4   645  351   92  622   84  192   18   72  265  311
0   649  456  507  355  210  158  295   12  609  460

[65 rows x 10 columns]
```
I also added a test for when we have enough partitions that we have multiple col + row partitions (so that case with additional columns):
```python
In [1]: from modin.pandas.test.utils import eval_general
   ...: from modin.config import NPartitions; NPartitions.put(24)
   ...: import pandas
   ...: import numpy as np
   ...: import modin.pandas as pd
   ...:
   ...:
   ...: # Next we test when the final df will have multiple row partitions.
   ...: data = np.random.choice(6500, 6500, replace=False).reshape((65, 100))
   ...: modin_df = pd.concat([pd.DataFrame(row).T for row in data]).reset_index(drop=True)
   ...: pandas_df = pandas.DataFrame(data)
   ...: print("Partition Shape:", modin_df._query_compiler._modin_frame._partitions.shape)
   ...: eval_general(modin_df, pandas_df, lambda df: df.sort_values(by=0)) # fails
UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
UserWarning: Ray execution environment not yet initialized. Initializing...
To remove this warning, run the following python code before doing dataframe operations:

    import ray
    ray.init(runtime_env={'env_vars': {'__MODIN_AUTOIMPORT_PANDAS__': '1'}})

UserWarning: Distributing <class 'numpy.ndarray'> object. This may take some time.
Partition Shape: (22, 4)
```
I've also changed the tests to test these scenarios!